src/api-service/__app__/onefuzzlib/secrets.py

```diff
@@ -0,0 +1,70 @@
+#!/usr/bin/env python
+#
+# Copyright (c) Microsoft Corporation.
+# Licensed under the MIT License.
+
+
+from typing import Tuple, cast
+from urllib.parse import urlparse
+from uuid import uuid4
+
+from azure.keyvault.secrets import KeyVaultSecret
+from onefuzztypes.models import SecretAddress, SecretData
+from pydantic import BaseModel
+
+from .azure.creds import get_instance_name, get_keyvault_client
+
+
+def save_to_keyvault(secret_data: SecretData) -> None:
+    if isinstance(secret_data.secret, SecretAddress):
+        return
+
+    secret_name = str(uuid4())
+    if isinstance(secret_data.secret, str):
+        secret_value = secret_data.secret
+    elif isinstance(secret_data.secret, BaseModel):
+        secret_value = secret_data.secret.json()
+    else:
+        raise Exception("invalid secret data")
+
+    kv = store_in_keyvault(get_keyvault_address(), secret_name, secret_value)
+    secret_data.secret = SecretAddress(url=kv.id)
+
+
+def get_secret_string_value(self: SecretData[str]) -> str:
```

Comment:
This really seems like it should be a getter on the `SecretData` class. Is it not because you want it only to be defined for the concrete type `SecretData[str]`, vs. the generic `SecretData[T]`?

Reply:
agree but I did not want to add a dependency on the keyvault library to the onefuzztypes.
I tried extending the type here but i struggled to find the right syntax (if you have a suggestion let me know)

Reply:
Does the `SecretData` class even make sense without keyvault methods enabled? It seems like the semantics basically require it. Could we just have it NOT live in the `pytypes` package?

Reply:
Secret data is used to post data to the service and show the information to the user inthe console. The keyvault api is only needed by the backend to save and retrieve the secret
We still need to reference SecretData in the notification template classes defined in `pytypes`.

Reply:
Got it. This is easy to understand, so let's keep it, and we can maybe revisit later. We can imagine defining a subclass of `SecretData` that is local to `onefuzzlib`, and which extends the base class to add these "live" methods that interact with keyvault, but I see that as more of a syntactic difference.